src/unix/udp.c

```diff
@@ -248,23 +279,43 @@ static void uv__udp_recvmsg(uv_udp_t* handle) {
     h.msg_iov = (void*) &buf;
     h.msg_iovlen = 1;
 
+    if(handle->udp_flags & UV_UDP_PKTINFO) {
+      h.msg_control = (void *)cmbuf;
+      h.msg_controllen = sizeof(cmbuf);
+    }
+
     do {
       nread = recvmsg(handle->io_watcher.fd, &h, 0);
     }
     while (nread == -1 && errno == EINTR);
 
     if (nread == -1) {
       if (errno == EAGAIN || errno == EWOULDBLOCK)
-        handle->recv_cb(handle, 0, &buf, NULL, 0);
+        handle->recv_cb(handle, 0, &buf, NULL, NULL, 0);
       else
-        handle->recv_cb(handle, UV__ERR(errno), &buf, NULL, 0);
+        handle->recv_cb(handle, UV__ERR(errno), &buf, NULL, NULL, 0);
     }
     else {
       flags = 0;
       if (h.msg_flags & MSG_TRUNC)
         flags |= UV_UDP_PARTIAL;
 
-      handle->recv_cb(handle, nread, &buf, (const struct sockaddr*) &peer, flags);
+      if (handle->udp_flags & UV_UDP_PKTINFO) {
+        struct cmsghdr* cmsg = CMSG_FIRSTHDR(&h);
+        for (; cmsg != NULL; cmsg = CMSG_NXTHDR(&h, cmsg)) {
+          /* Ignore all non-pktinfo messages */
+          if(cmsg->cmsg_level != IPPROTO_IP || cmsg->cmsg_type != IP_PKTINFO) {
+            continue;
+          }
+          struct in_pktinfo *pi = (struct in_pktinfo *)CMSG_DATA(cmsg);
+          struct sockaddr_in dst = {0};
+          dst.sin_family = AF_INET;
+          dst.sin_addr.s_addr = pi->ipi_spec_dst.s_addr;
```

Comment:
We do not fill the port in `dst_peer` in recv_cb. 
I think it is omitted since we will know the listening port even if udp socket listens on 0.0.0.0
But not sure whether it is fine as an api design since `sockaddr` include port too.

Reply:
The "pktinfo" struct is:

```
struct in_pktinfo {
  unsigned int ipi_ifindex; /* Interface index */
  struct in_addr ipi_spec_dst; /* Local address */
  struct in_addr ipi_addr; /* Header Destination address */
}
```

It doesn't contain any port info, that's why we don't fill the port in `dst` here.